### DISCUSSION TITLE
[RFC] Attachment component

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/x#159] [RFC]
## Summary

新增 Attachment 组件用于附件上传、关联能力支持。

## Motivation

上传管理中有多种方式，对于纯文件上传使用 antd Upload 即可：

<img width="812" alt="截屏2024-10-16 11 30 28" src="https://github.com/user-attachments/assets/b6a171fa-b406-4680-98b3-1c86aa107b58">

<img width="1177" alt="截屏2024-10-16 11 31 02" src="https://github.com/user-attachments/assets/136178b7-0142-4f3a-b675-90efd7ab27fa">

对于拖拽以及附件管理，则需要封装支持：

* 有实体时默认为上传提示面板，拖拽后转换为附件列表
* 无实体时，拖拽则直接显示模糊效果

<img width="1162" alt="截屏2024-10-16 11 32 17" src="https://github.com/user-attachments/assets/23aaa71b-1125-4fb2-b4ea-9b41e20a2cd0">

<img width="705" alt="截屏2024-10-16 11 32 39" src="https://github.com/user-attachments/assets/c40850bb-701c-42d6-9a72-165c736794e3">

<img width="699" alt="截屏2024-10-16 11 33 00" src="https://github.com/user-attachments/assets/b64a5b6c-1610-43ec-b72a-f890520c3581">

<img width="1010" alt="截屏2024-10-16 11 33 23" src="https://github.com/user-attachments/assets/40dd3775-5808-4abc-9db3-989da13d9701">

<img width="991" alt="截屏2024-10-16 11 34 11" src="https://github.com/user-attachments/assets/ab689353-3d15-4ee8-a267-3dd93acf555d">

### 无实体做纯全屏 Upload 监听

```tsx
<Attachment getDropContainer{() => document.body}>
  <Button />
</Attachement>
```

### 有实体做面板使用

```tsx
<Attachment
  // 无附件时展示配置
  placeholder={{
    icon: <CloudUploadOutlined />,
    title: "拖拽上传",
    description: "支持 pdf, txt, jpg 格式"
  }}

  // 有附件时展示配置
  overflow="scroll | wrap | responsive"
  
  // 数据
  data={data}
  onChange={onChange}
/>
```

|  属性         | 说明               | 类型                    | 默认值 |
| ------------ | ------------------ | ----------------------- | ------ |
| children     | 子元素存在时作为 Upload 功能使用           | ReactElement               | -      |
| getDropContainer | 设置需要拖拽放置的目标，当鼠标拖拽文件时会提示用户到这个位置 | - |
| placeholder | 没有附件时展示的内容 | { icon, title, description } \| ReactElement | - |
| overflow | 超出时设置滚动效果 | scroll \| wrap \| responsive | responsive |
| classNames | 语义化自定义样式 | Record<SemanticType, string> | - |
| styles | 语义化自定义样式 | Record<SemanticType, CSSProps> | - |
| data | 数据列表 | antdFile[] | - | 
| onChange | 数据列表变化回调 | (files: antdFile[]) => void | - | 

```tsx
type SemanticType = 'placeholder' | 'item' | 'list';
```

### Sender 配合示例

```tsx
const header = (
  <Sender.Header open={open}>
    <Attachment  data={data} onChange={setData} />
  </Sender.Header>
);

<Sender
  header={header}
  onSubmit={(text) => {
    agent.request({
      message: { text, data },
    });
    setData([]);
    setOpen(false);
  }}
/>
```

## What's more

关联外部站点、文档等功能，待设计。

**Comment by YumoImer** [comment]
data -> items

**Reply by zombieJ** [reply]
一下子让我纠结了，antd 的规范里和数据相关的新的都叫 data，和渲染相关的叫 items。这里看起来两者都可以。
